File: LeetCode/src/designclass/LRUCache.java
package designclass;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

public class LRUCache {
    public static void main(String[] args) {
        LRUCacheByLinkedList cache1 = new LRUCacheByLinkedList(2);
        cache1.put(1, 1);
        cache1.put(2, 2);
        System.out.println(cache1.get(1)); // Output: 1
        cache1.put(3, 3);
        System.out.println(cache1.get(2)); // Output: -1
        cache1.put(4, 4);
        System.out.println(cache1.get(1)); // Output: -1
        System.out.println(cache1.get(3)); // Output: 3
        System.out.println(cache1.get(4)); // Output: 4

        LRUCacheByLinkedHashMap cache2 = new LRUCacheByLinkedHashMap(2);
        cache2.put(1, 1);
        cache2.put(2, 2);
        System.out.println(cache2.get(1)); // Output: 1
        cache2.put(3, 3);
        System.out.println(cache2.get(2)); // Output: -1
        cache2.put(4, 4);
        System.out.println(cache2.get(1)); // Output: -1
        System.out.println(cache2.get(3)); // Output: 3
        System.out.println(cache2.get(4)); // Output: 4
    }

}
class LRUCacheByLinkedList {
    private int capacity;
    private Map<Integer, ListNode> map;
    private ListNode head;
    private ListNode tail;

    public LRUCacheByLinkedList(int capacity) {
        this.capacity = capacity;
        map = new HashMap<>();
        head = new ListNode(-1,-1);
        tail = new ListNode(-1, -1);
        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        if (!map.containsKey(key)) {
            return -1;
        }

        ListNode node = map.get(key);
        remove(node);
        add(node);
        return node.val;
    }

    public void put(int key, int value) {
        if (map.containsKey(key)) {
            ListNode oldNode = map.get(key);
            remove(oldNode);
        }

        ListNode node = new ListNode(key, value);
        map.put(key, node);
        add(node);

        if (map.size() > capacity) {
            ListNode removeNode = head.next;
            remove(removeNode);
            map.remove(removeNode.key);
        }
    }

    private void add(ListNode node) {
        ListNode tailPrev = tail.prev;
        tailPrev.next = node;
        node.prev = tailPrev;
        node.next = tail;
        tail.prev = node;
    }

    private void remove(ListNode node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }
}

class ListNode {
    int key;
    int val;
    ListNode prev;
    ListNode next;
    public ListNode(int key, int val) {
        this.key = key;
        this.val = val;
    }
}

class LRUCacheByLinkedHashMap {

    private int capacity;
    private LinkedHashMap<Integer, Integer> cache;

    public LRUCacheByLinkedHashMap(int capacity) {
        this.capacity = capacity;
        this.cache = new LinkedHashMap<Integer, Integer>(capacity, 0.75f, true) {
            @Override
            protected boolean removeEldestEntry(Map.Entry<Integer, Integer> eldest) {
                return size() > capacity;
            }
        };
    }

    public int get(int key) {
        return cache.getOrDefault(key, -1);
    }

    public void put(int key, int value) {
        cache.put(key, value);
    }
}

File: LeetCode_Python/designclass/LRUCache.py
class Node:
    def __init__(self, key: int, value: int):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.head = Node(0, 0)
        self.tail = Node(0, 0)
        self.head.next = self.tail
        self.tail.prev = self.head
    
    def remove(self, node: Node) -> None:
        node.prev.next = node.next
        node.next.prev = node.prev

    def add_to_tail(self, node: Node) -> None: 
        tail_prev = self.tail.prev
        tail_prev.next = node
        node.prev = tail_prev
        node.next = self.tail
        self.tail.prev = node

    def get(self, key: int) -> int:
        if key in self.cache:
            node = self.cache[key]
            self.remove(node)
            self.add_to_tail(node)
            return node.value
        return -1
        
    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            node.value = value
            self.remove(node)
            self.add_to_tail(node)
        else:
            new_node = Node(key, value)
            self.cache[key] = new_node
            self.add_to_tail(new_node)
            if len(self.cache) > self.capacity:
                lru = self.head.next
                self.remove(lru)
                del self.cache[lru.key]

if __name__ == '__main__':
    cache = LRUCache(2)
    cache.put(1, 1)
    cache.put(2, 2)
    print("get(1): Expected output 1, Actual output:", cache.get(1))
    cache.put(3, 3)  # This should evict key 1 (if we use head insertion) or key 2 (if tail insertion)
    print("get(2): Expected output -1 (if key 2 is evicted) or 2 otherwise, Actual output:", cache.get(2))
    cache.put(4, 4)
    print("get(1): Expected output -1, Actual output:", cache.get(1))
    print("get(3): Expected output 3, Actual output:", cache.get(3))
    print("get(4): Expected output 4, Actual output:", cache.get(4))